Fixed #26561 -- Fixed unusably slow SelectBox performance with large lists

dsanders11 opened the pull request:
Discussion on [ticket 26561](https://code.djangoproject.com/ticket/26561). Patch to at least make it usable with large lists in 1.10, with the long-term rewrite on [ticket 15220](https://code.djangoproject.com/ticket/15220) still to come.

charettes commented:
Using a `jQuery` instead of raw JavaScript doesn't serve your optimization purposes very well. I believe this can be achieved without using it.

charettes commented on django/contrib/admin/static/admin/js/SelectBox.js:
```diff
@@ -125,11 +121,8 @@
-            for (var i = 0; i < box.options.length; i++) {
-                box.options[i].selected = 'selected';
-            }
+            $('#' + id + ' option').prop('selected', 'selected');
```
Simply caching `box.options` and `box.options.length` in local variables should be enough here.

charettes commented on django/contrib/admin/static/admin/js/SelectBox.js:
```diff
@@ -79,30 +79,26 @@
+                    SelectBox.delete_from_cache(from, value);
                 }
-            }
+            });
```
Caching `boxOptions.length` in a local variable and `option.value`, `option.text` per iteration should do.

dsanders11 replied:
The main benefit here is the 'option:selected' selector which avoids walking all options and doing a check if they're selected in favor of only walking the ones which are selected.

charettes replied:
The thing is there's no such thing as a `:selected` selector, it's a [`jQuery`](https://api.jquery.com/selected-selector/) extension. You end up iterating over you collection of `option`s in both cases.

dsanders11 replied:
Good point. Since I needed to use jQuery for the cross-browser compatible emptying of the `select` I figured I'd move more of the code to jQuery (since there is still ticket 15220 which aims to move all of selectfilter to jQuery) but it looks like the performance benefit is negligible. Had too much faith that jQuery could be smarter about some of the operations.

charettes commented on django/contrib/admin/static/admin/js/SelectBox.js:
```diff
@@ -79,30 +79,26 @@
+                    SelectBox.delete_from_cache(from, value);
                 }
-            }
+            });
```
Ditto.

charettes commented on django/contrib/admin/static/admin/js/SelectBox.js:
```diff
@@ -37,11 +39,13 @@
-                    if (node.text.toLowerCase().indexOf(token) === -1) {
+                    if (node_text.indexOf(token) === -1) {
                         node.displayed = 0;
+                        break;  // Once the first token isn't found we're done
```
Nice catch here, this should make a big difference.

charettes commented on django/contrib/admin/static/admin/js/SelectBox.js:
```diff
@@ -57,11 +61,7 @@
-                cache[i] = cache[i + 1];
-            }
-            cache.length--;
+            cache.splice(delete_index, 1);
```
Ouch, this code is really showing its age.

dsanders11 commented:
@charettes, the bulk of the speed boost comes from the changes to `redisplay`, as that was the biggest slowdown culprit. Unfortunately there's an [IE bug](https://support.microsoft.com/en-us/kb/276228) with setting innerHTML on a select (which I originally did to clear it out), so I figured `$.empty` would handle the cross-browser compatibility problems there.

dsanders11 commented:
@charettes, I've redone the commit to address your comments. Now the only use of jQuery is for $.empty. If there's a quick and easy cross-browser compatible alternative we can drop it all together.

The main performance gains (in order of impact) are from:
* Refactoring redisplay to rebuild the select in a string instead of in DOM
* Using splice in delete_from_cache
* Stopping filter for an option after the first token miss

The rest of the changes, using local variables to avoid re-executing code inside of loops, offer a very small performance impact compared to the above and could be dropped if there is a desire to keep the patch smaller and more focused.

charettes commented:
Buildbot, test on selenium.

dsanders11 commented:
The selenium test results are a bit confusing. Looks like there was at least one test mishap (address already in use during setUp), a couple failures for JS i18n which this shouldn't affect at all, and then the seemingly relevant failures are for failing to activate the 'select' icon when clicking inside the from box, but that should be entirely handled by SelectFilter2.js, since it binds a change handler and refreshes the icons on change, never calling SelectBox.js during that process.

¯\\\_(ツ)_/¯

dsanders11 commented:
@charettes, I found the cause of the selenium test failure, it was that the event listeners for click events on the `select` boxes were being lost since now the entire select is being recreated. I fixed the issue by moving the binding up further in the DOM and checking that the events were for `select`/`option`.

I also added another performance fix as another commit (slight scope creep) because I found a good fix for the very slow (I was seeing 5 seconds for a list of 20k options) `SelectFilter.refresh_icons` on Chrome and Safari. It's a little clever trick that momentarily marks the `select` as required and asks the browser if it's valid via the `:valid` to see if there are any selected options. This was much, much faster in my testing. The `is(':valid')` could be changed to `checkValidity()` if preferred, but the latter fires an event according to the documentation so I thought a quick check of the CSS selector would be cleaner.

timgraham commented:
merged in fb68674ea4b05280e3deb2ed23602038755092bc, thanks!

charettes commented:
Thanks for your work on this David.

dsanders11 commented:
@charettes, no problem. Happy it was able to make it in for 1.10, that'll make some people over at [django-newsletter](https://github.com/dokterbob/django-newsletter/) happy.

@timgraham, of course right after you merged this I spotted one line that could be improved, but it's not a big deal. [Line 166 in SelectBox.js](https://github.com/django/django/pull/6537/files#diff-62f31c141a8122aaaa321291c99606a9R166) would be better moved outside of that else so that it is always run even if `is(':valid')` throws an error (which it will on IE 9). If an error throws there the `required` attribute will be left on the field, but I don't think it's a big deal because if the error occurs then the browser doesn't support HTML5's required to begin with, so in all likelihood it will entirely ignore the required attribute. Just a hygiene thing with it being left there inadvertently.
